flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1169,25 +1169,28 @@ class ScalarExprLowering {
 
   fir::ExtendedValue genArrayLit(
       const Fortran::evaluate::Constant<Fortran::evaluate::SomeDerived> &con) {
+    auto loc = getLoc();
+    auto idxTy = builder.getIndexType();
+    auto size = Fortran::evaluate::GetSize(con.shape());
+    fir::SequenceType::Shape shape(con.shape().begin(), con.shape().end());
+    auto eleTy = converter.genType(con.GetType().GetDerivedTypeSpec());
+    auto arrayTy = fir::SequenceType::get(shape, eleTy);
+    mlir::Value array = builder.create<fir::UndefOp>(loc, arrayTy);
+    if (size == 0)
+      return fir::ArrayBoxValue{array, {}, {}};
```

Comment:
The assumption is that this return box is never used.  That seems to be the case in practice, because an instance of this code has been in use for some time.  That said, we do still have unanalyzed bugs.  However, at the cost of generating dead fir that will later be removed for these rare cases, I will shift to generating bounds constants all the time.  Thanks for pointing this out.